cpp/src/arrow/util/async_generator.h

```diff
@@ -221,6 +221,9 @@ class MappingGenerator {
       bool should_trigger;
       {
         auto guard = state->mutex.Lock();
+        // A MappedCallback may have purged or be purging the queue;
+        // we shouldn't do anything here.
+        if (state->finished) return;
```

Comment:
Shoot, I kind of forgot about this but I have this fix in my CSV streaming PR I kind of half-abandoned because it was too complex and I needed to work on other things.  Can you grab the QueuingMapFailStress test case from https://github.com/apache/arrow/pull/10795/files#diff-aa924daf1e09b13ad031b70e40eeff3b4d4b817b2a32981bb0c0fd2e5313326cR500

I like your fix better anyways.  Instead of returning early I just added more flags and ifs which was a more complex version of the same thing.

Reply:
That test case uses an overload of MakeMappedGenerator which doesn't exist in the fork (it takes a bool as the third argument) - what was the purpose there?

Reply:
You can ignore the third parameter.  Odd that the overload doesn't exist :confused:.  It was basically to toggle between a newly added BasicMappingGenerator (which pulls from source reentrantly) and the regular mapping generator (which does not pull from the source reentrantly but does allow the map function to run in parallel).  I think the test is always using `true` which maps to what we currently have.  So just drop the `true` and you should be ok.

Reply:
Ah I figured it would be something like that, I added the test here.